community/kernel/src/main/java/org/neo4j/kernel/impl/locking/CombinedTracer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.locking;
+
+import java.util.Arrays;
+
+import org.neo4j.storageengine.api.lock.ResourceType;
+
+final class CombinedTracer implements Locks.Tracer
+{
+    private final Locks.Tracer[] tracers;
+
+    public CombinedTracer( Locks.Tracer... tracers )
+    {
+        this.tracers = tracers;
+    }
+
+    @Override
+    public Locks.WaitEvent waitForLock( ResourceType resourceType, long... resourceIds )
+    {
+        Locks.WaitEvent[] events = new Locks.WaitEvent[tracers.length];
+        for ( int i = 0; i < events.length; i++ )
+        {
+            events[i] = tracers[i].waitForLock( resourceType, resourceIds );
+        }
+        return new CombinedEvent( events );
+    }
+
+    @Override
+    public Locks.Tracer combine( Locks.Tracer tracer )
+    {
+        if ( tracer == NONE )
+        {
+            return this;
+        }
+        Locks.Tracer[] tracers;
+        if ( tracer instanceof CombinedTracer )
+        {
+            Locks.Tracer[] those = ((CombinedTracer) tracer).tracers;
+            tracers = Arrays.copyOf( this.tracers, this.tracers.length + those.length );
+            for ( int i = 0; i < those.length; i++ )
+            {
+                tracers[this.tracers.length + i] = those[i];
+            }
+        }
+        else
+        {
+            tracers = Arrays.copyOf( this.tracers, this.tracers.length + 1 );
+            tracers[this.tracers.length] = tracer;
+        }
```

Comment:
if by `n` you mean the length of the `tracers` array, I'd anticipate the size to be `3` if this code ever executes, typically I'd expect the size to be `2` in instances of `CombinedTracer`, do you think we should make it tree structured instead of flat? (since that would optimize for the expected size of 2)